WalletWasabi/TorSocks5/TorProcessManager.cs

```diff
@@ -32,66 +35,87 @@ public async Task StartAsync()
+					Logger.LogInfo<TorProcessManager>($"Extracted {dataZip} to {torDir}.");
 
 					if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
 					{
-						try
+						string torWinZip = Path.Combine(torDaemonsDir, "tor-win32.zip");
+						await IoHelpers.BetterExtractZipToDirectoryAsync(torWinZip, torDir);
+						Logger.LogInfo<TorProcessManager>($"Extracted {torWinZip} to {torDir}.");
+					}
+					else // Linux or OSX
+					{
+						if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
 						{
-							ZipFile.ExtractToDirectory(Path.Combine(torDaemonsDir, "tor-win32.zip"), torDir);
+							string torLinuxZip = Path.Combine(torDaemonsDir, "tor-linux64.zip");
+							await IoHelpers.BetterExtractZipToDirectoryAsync(torLinuxZip, torDir);
+							Logger.LogInfo<TorProcessManager>($"Extracted {torLinuxZip} to {torDir}.");
 						}
-						catch (UnauthorizedAccessException)
+						else // OSX
 						{
-							await Task.Delay(100);
-							ZipFile.ExtractToDirectory(Path.Combine(torDaemonsDir, "tor-win32.zip"), torDir);
+							string torOsxZip = Path.Combine(torDaemonsDir, "tor-osx64.zip");
+							await IoHelpers.BetterExtractZipToDirectoryAsync(torOsxZip, torDir);
+							Logger.LogInfo<TorProcessManager>($"Extracted {torOsxZip} to {torDir}.");
 						}
+
+						// Make sure there's sufficient permission.
+						string chmodTorDirCmd = $"chmod -R 777 {torDir}";
+						EnvironmentHelpers.ShellExec(chmodTorDirCmd);
+						Logger.LogInfo<TorProcessManager>($"Shell command executed: {chmodTorDirCmd}.");
 					}
-					else
-					{
-						throw new NotImplementedException();
-					}
+				}
+				else
+				{
+					Logger.LogInfo<TorProcessManager>($"Tor instance found at {torPath}.");
 				}
 
-				var torProcessStartInfo = new ProcessStartInfo(torPath)
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
 				{
-					Arguments = $"SOCKSPort {TorSocks5EndPoint.Port}",
-					UseShellExecute = false,
-					CreateNoWindow = true,
-					RedirectStandardOutput = true
-				};
-				Process torProcess = Process.Start(torProcessStartInfo);
+					var torProcessStartInfo = new ProcessStartInfo(torPath)
+					{
+						Arguments = $"SOCKSPort {TorSocks5EndPoint.Port}",
+						UseShellExecute = false,
+						CreateNoWindow = true,
+						RedirectStandardOutput = true
+					};
+					Process.Start(torProcessStartInfo);
+					Logger.LogInfo<TorProcessManager>($"Starting Tor process with Process.Start.");
+				}
+				else // Linux and OSX
+				{
+					string runTorCmd = $"LD_LIBRARY_PATH=$LD_LIBRARY_PATH:={torDir}/Tor && export LD_LIBRARY_PATH && cd {torDir}/Tor && ./tor SOCKSPort {TorSocks5EndPoint.Port}";
+					EnvironmentHelpers.ShellExec(runTorCmd, false);
+					Logger.LogInfo<TorProcessManager>($"Started Tor process with shell command: {runTorCmd}.");
+				}
 
-				await Task.Delay(1000);
+				await Task.Delay(1000).ConfigureAwait(false); // dotnet brainfart, ConfigureAwait(false) IS NEEDED HERE otherwise (only on) Manjuro Linux fails, WTF?!!
```

Comment:
@lontivero Something REALLY fucked up is happenning.  
I had to add `.ConfigureAwait(false);` here, otherwise it gets stuck on Manjuro Linux (Manjuro is based on Arch Linux.)

Reply:
I opened a stackoverflow discussion: https://stackoverflow.com/questions/52093132/net-core-2-1-manjuro-linux-fails-without-configureawaitfalse